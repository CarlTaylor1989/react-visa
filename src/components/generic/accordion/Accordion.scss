@import '../../../styles/mixins';
@import '../../../styles/variables';

$dot-diameter: 6px;

.accordionItem {
  padding-bottom: 10px;

  .selectedNumberIndicatorWrapper {
    @include align-items(center);
    display: flex;
    position: absolute;
    right: 45px;
    top: 14px;

    .selectedNumberIndicator {
      background-color: $colour-white;
      border-radius: ($dot-diameter / 2);
      display: block;
      height: $dot-diameter;
      margin-left: 6px;
      width: $dot-diameter;
    }
  }
}
