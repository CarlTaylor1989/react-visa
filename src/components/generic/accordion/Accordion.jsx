import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Block from './Block';

import './Accordion.scss';

class Accordion extends Component {
  constructor(props) {
    super(props);
    const { open, children } = this.props;
    this.state = {
      prev: open,
      flags: new Array(React.Children.count(children)).fill(false).map(
        (val, index) => (index === open) // creates a boolean array with length set to the number of children and all items set to false except item at index props.open set to true
      )
    };
    this.onAccordionChange = this.onAccordionChange.bind(this);
  }

  onAccordionChange(index) {
    const newState = Object.assign({}, this.state);
    const { prev, flags } = this.state;
    if (
      prev !== index
      && prev !== null
      && flags[prev]
    ) {
      newState.flags[prev] = false;
    }
    newState.flags[index] = !flags[index];
    newState.prev = index;
    this.setState({ ...newState });
  }

  createBlocks(children) {
    const { flags } = this.state;

    return React.Children.map(children, (child, index) => (
      <Block
        title={child.props.title}
        className={child.props['data-custom-class']}
        selected={child.props['data-selected']}
        hasProduct={child.props['data-has-product']}
        isOpen={flags[index]}
        onToggle={() => this.onAccordionChange(index)}
      >
        {child}
      </Block>
    ));
  }

  render() {
    const { children, className } = this.props;
    const blocks = this.createBlocks(children);

    return (
      <div className={className}>
        {blocks}
      </div>
    );
  }
}

Accordion.propTypes = {
  children: PropTypes.node.isRequired,
  open: PropTypes.number.isRequired,
  className: PropTypes.string
};

Accordion.defaultProps = {
  className: ''
};

export default Accordion;
