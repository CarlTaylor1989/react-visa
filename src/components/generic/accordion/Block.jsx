import React from 'react';
import PropTypes from 'prop-types';
// eslint-disable-next-line import/no-unresolved
import Collapse from '@kunukn/react-collapse';

import './Accordion.scss';

const Block = ({
  isOpen, title, onToggle, children, className, selected, hasProduct
}) => (
  <div className="accordionItem">
    <button
      className={`toggleButton ${className}`}
      type="button"
      onClick={() => (hasProduct ? onToggle() : null)}
    >
      <span className="accordionItemTitle">
        {title}
      </span>
      <div className="selectedNumberIndicatorWrapper">
        {selected.map(sel => <span key={sel} className="selectedNumberIndicator" />)}
      </div>
      {hasProduct && <span className={`icon-down ${isOpen ? 'is-open' : ''}`} /> }
    </button>
    <Collapse isOpen={isOpen}>{children}</Collapse>
    <span>{children.length}</span>
  </div>
);

Block.propTypes = {
  isOpen: PropTypes.bool.isRequired,
  title: PropTypes.string.isRequired,
  onToggle: PropTypes.func.isRequired,
  children: PropTypes.node.isRequired,
  selected: PropTypes.array,
  className: PropTypes.string,
  hasProduct: PropTypes.bool.isRequired
};

Block.defaultProps = {
  className: '',
  selected: []
};

export default Block;
