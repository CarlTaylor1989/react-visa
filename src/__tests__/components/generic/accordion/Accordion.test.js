/* global toJson */
import React from 'react'; // eslint-disable-line no-unused-vars
import Accordion from '../../../../components/generic/accordion/Accordion';
import Block from '../../../../components/generic/accordion/Block';

let mockProps;
beforeEach(() => {
  mockProps = {
    open: 1,
    children: ([
      <div title="first_item" data-has-product>
        <p>Paragraph of text for first item.</p>
      </div>,
      <div title="second_item" data-has-product>
        <p>Paragraph of text for second item.</p>
      </div>,
      <div title="third_item" data-has-product={false} />
    ])
  };
});

test('renders the Accordion component', () => {
  const wrapper = shallow(<Accordion {...mockProps} />);
  expect(toJson(wrapper, { noKey: false })).toMatchSnapshot();
});

test('sets initial state', () => {
  const wrapper = shallow(<Accordion {...mockProps} />);
  const finalState = {
    prev: 1,
    flags: [false, true, false]
  };
  expect(wrapper.instance().state).toStrictEqual(finalState);
});

test('sets initial state without product', () => {
  const wrapper = shallow(<Accordion {...mockProps} />);
  const finalState = {
    prev: 1,
    flags: [false, true, false]
  };
  expect(wrapper.instance().state).toStrictEqual(finalState);
});

describe('method onAccordionChange()', () => {
  test('sets the previous flag to false', () => {
    // setup
    const wrapper = shallow(<Accordion {...mockProps} />);
    wrapper.instance().setState({
      prev: 0,
      flags: [true, false, false]
    });
    const finalState = {
      prev: 2,
      flags: [false, false, true]
    };

    // function under test
    wrapper.instance().onAccordionChange(2);

    // expectation
    expect(wrapper.instance().state).toStrictEqual(finalState);
  });

  test('sets the current flag to true', () => {
    // setup
    const wrapper = shallow(<Accordion {...mockProps} />);
    wrapper.instance().setState({
      prev: 0,
      flags: [false, false, false]
    });
    const finalState = {
      prev: 0,
      flags: [true, false, false]
    };

    // function under test
    wrapper.instance().onAccordionChange(0);

    // expectation
    expect(wrapper.instance().state).toStrictEqual(finalState);
  });
});

test('method accordion block toggled', () => {
  const wrapper = shallow(<Accordion {...mockProps} />);
  wrapper.instance().onAccordionChange = jest.fn();

  // function under test
  wrapper.find(Block).at(0).props().onToggle();

  // expectation
  expect(wrapper.instance().onAccordionChange).toHaveBeenCalledWith(0);
});
