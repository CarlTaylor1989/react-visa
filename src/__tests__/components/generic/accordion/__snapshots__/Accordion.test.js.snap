// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the Accordion component 1`] = `
<div
  className=""
>
  <Block
    className=""
    hasProduct={true}
    isOpen={false}
    key=".0"
    onToggle={[Function]}
    selected={Array []}
    title="first_item"
  >
    <div
      data-has-product={true}
      title="first_item"
    >
      <p>
        Paragraph of text for first item.
      </p>
    </div>
  </Block>
  <Block
    className=""
    hasProduct={true}
    isOpen={true}
    key=".1"
    onToggle={[Function]}
    selected={Array []}
    title="second_item"
  >
    <div
      data-has-product={true}
      title="second_item"
    >
      <p>
        Paragraph of text for second item.
      </p>
    </div>
  </Block>
  <Block
    className=""
    hasProduct={false}
    isOpen={false}
    key=".2"
    onToggle={[Function]}
    selected={Array []}
    title="third_item"
  >
    <div
      data-has-product={false}
      title="third_item"
    />
  </Block>
</div>
`;
